### PR TITLE
Fixes #16284 - Move Turbolinks to npm

### PR DESCRIPTION
**Pull request by shlomizadok**
A few notes:
- Development & test require coffeescript gem (was apparently part of turbolinks gem)
- Tried to move `jquery.turbolinks` to npm as well, alas it didn't work well :(

### CODE REVIEW

File: webpack/assets/javascripts/bundle.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ require('expose?JsDiff!diff');
 require('./bundle_flot');
 require('./bundle_select2');
 require('./bundle_datatables');
+var Turbolinks = require('turbolinks');
```
Comment (tbrisker) [inline webpack/assets/javascripts/bundle.js]: Is the `Turbolinks` variable needed in any of the older js? if so, please use the expose loader (see examples a few lines above)

**Comment by tbrisker** [comment]
A couple of questions about the notes:
- Why do we need coffeescript? do we actually use it anywhere?
- What didn't work with jquery.turbolinks? what is the difference between that and this? do we actually use it at all?

File: webpack/assets/javascripts/bundle.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gailsteiger) [reply webpack/assets/javascripts/bundle.js]: This code creates a global. This is a very bad practice for javascript. We don't allow that, except in the case Tomer mentioned below. If the expose loader doesn't work for you, please add Turbolinks as a property of the window.tfm object.

**Comment by gailsteiger** [comment]
Why is coffeescript required?

**Comment by gailsteiger** [comment]
As a general comment, I would always prefer a npm module to a jquery plugin.

**Comment by shlomizadok** [comment]
1. Coffee-Script, apparently a dependency add somewhere in Rails. We do not use it.
2. You can't require jquery.turbolinks, after installing the npm (https://www.npmjs.com/package/jquery.turbolinks) - weirdly it just adds the gem files

**Comment by dLobatog** [comment]
Setting this to 'waiting on contributor' as HostJS integration tests are failing

**Comment by shlomizadok** [comment]
Weird. It passes locally :/

**Comment by shlomizadok** [comment]
So here's the deal:
We currently use `turbolinks ~>2.5`  gem. Alas, the turbolinks npm version is 5.0.0

Same failures will occur if we'll upgrade our gem to `turbolinks ~>5.0`

We can either:
1. Ditch turbolinks (I am unsure how valuable it is anyhow) - I change this pull request to "Remove turbolinks"
2. Keep current gem and close this pull request

Please add your reaction:
- If in favor of removing turbolinks, react with :+1: 
- If in favor of keeping turbolinks, react with :-1:

**Comment by gailsteiger** [comment]
I would be surprised if turbo links significantly improves performance.
I worry also that not refreshing the page completely in these cases put us at risk for other problems, including memory leaks. Also, if anyone is still binding event handlers to window, the handlers will remain attached after the 'navigation'

**Comment by ohadlevy** [comment]
I don't think we should remove it just because, if you have good arguments (with measurements in production and development environments, then I'm open to improvements..)

**Comment by shlomizadok** [comment]
@ohadlevy et. all, I may have rushed to rule out turbolinks and anyhow it should be discussed on a separate thread.
Currently, the change to npm includes:
- Turbolinks version 5.0.0 (we currently run 2.3.5)
- Adding `coffee-rails` gem as a dependency, as it is required by turbolinks ~> 5.0
- Hosts#edit - remove turbolinks as interface tab won't load (would love to hear another solution)
- Turbolinks.jquery is removed as it is not used

**Comment by dLobatog** [comment]
Many things do not work with Turbolinks 5 (a complete rewrite) and Foreman - including the dashboard widgets, all multi-select stuff, and others. 

I'd be so down to use it, as it finally allows us to work in a more similar way to pjax (e.g: we can just keep the top bar model always painted in the browser with data-turbolinks-permanent), but it would require some more work. 

I'm not sure if we can move Turbolinks to npm without upgrading though as I don't see our version published in npmjs.org, so if you want to use this PR to do everything related to the upgrade. I'd be happy to review. @shlomizadok

**Comment by guilleiguaran** [comment]
Hi rails core member here,

coffee-rails is a dependency only for old versions of turbolinks (now called [turbolinks-classic](https://github.com/turbolinks/turbolinks-classic)), starting from version 5.0 the coffee files are compiled to js before of distribution to rubygems making coffee-rails a dependency only during development.

I think turbolinks-classic can be published to npm but it would require having the ruby lib anyway (there are some functionalities [in ruby](https://github.com/turbolinks/turbolinks-classic/tree/master/lib/turbolinks)), in contrast turbolinks 5 is a js-only lib and integration is done through [turbolinks gem](https://github.com/turbolinks/turbolinks-rails)

I'll try to check this in detail during the next days.

**Pull request closed without merging**